flake8_trio/visitors/visitor91x.py

```diff
@@ -54,148 +106,232 @@ def __init__(self, *args: Any, **kwargs: Any):
+        self.save_state(
+            node,
+            "has_yield",
+            "safe_decorator",
+            "async_function",
+            "uncheckpointed_statements",
+            "loop_state",
+            "try_state",
+            copy=True,
+        )
+        self.uncheckpointed_statements = set()
+        self.has_yield = self.safe_decorator = False
+        self.loop_state = LoopState()
 
-        # disable checks in asynccontextmanagers by saying the function isn't async
-        self.async_function = not fnmatch_qualified_name(
-            node.decorator_list, *self.options.no_checkpoint_warning_decorators
+        self.async_function = (
+            node.asynchronous is not None
+            and not fnmatch_qualified_name_cst(
+                node.decorators, *self.options.no_checkpoint_warning_decorators
+            )
         )
         if not self.async_function:
             return
 
+        pos = self.get_metadata(PositionProvider, node).start
         self.uncheckpointed_statements = {
-            Statement("function definition", node.lineno, node.col_offset)
+            Statement("function definition", pos.line, pos.column)
         }
 
-        self.generic_visit(node)
-
-        self.check_function_exit(node)
+    def leave_FunctionDef(
+        self, original_node: cst.FunctionDef, updated_node: cst.FunctionDef
+    ) -> cst.FunctionDef:
+        if self.async_function:
+            # updated_node does not have a Position, so we must send original_node
+            self.check_function_exit(original_node)
+        self.restore_state(original_node)
+        return updated_node
 
     # error if function exits or returns with uncheckpointed statements
-    def check_function_exit(self, node: ast.Return | ast.AsyncFunctionDef):
+    def check_function_exit(self, node: cst.Return | cst.FunctionDef):
         for statement in self.uncheckpointed_statements:
-            self.error(
-                node,
-                "return" if isinstance(node, ast.Return) else "exit",
-                statement,
-                error_code="TRIO911" if self.has_yield else "TRIO910",
-            )
+            self.error_91x(node, statement)
 
-    def visit_Return(self, node: ast.Return):
+    def leave_Return(
+        self, original_node: cst.Return, updated_node: cst.Return
+    ) -> cst.Return:
         if not self.async_function:
-            return
-        self.generic_visit(node)
-        self.check_function_exit(node)
-
+            return updated_node
+        self.check_function_exit(original_node)
         # avoid duplicate error messages
         self.uncheckpointed_statements = set()
 
-    # disregard checkpoints in nested function definitions
-    def visit_FunctionDef(self, node: ast.FunctionDef):
-        self.save_state(node)
-        self.set_state(self.default, copy=True)
+        return updated_node
 
-    # checkpoint functions
-    def visit_Await(self, node: ast.Await | ast.Raise):
+    def error_91x(
```

Comment:
added error helper with the introduction of artificial statements/errors - which is the new solution to not knowing the state of the beginning of loops before they've been fully parsed.

Reply:
I think you've got a comment in the code about this further down, but I think a similar short explanation here (1-2 lines?) would be nice.